### PR TITLE
For #10739 - Add support for the Storage Access API prompt

### PR DESCRIPTION
**Pull request by Mugurell**
| video | still |
| --- | --- |
| <video src="https://user-images.githubusercontent.com/11428869/146047209-8b56d687-9938-4db0-8c78-a278af47ee51.mp4" width="300" /> |  |

To get this prompt we need to change in `about:config`:
- `dom.storage_access.auto_grants` to `false`
(to disable Gecko automatically granting the permission)
- `network.cookie.cookieBehavior` to 5
- `network.cookies.cookieBehavior.pbmode` to 5
(to enable cross origin cookie blocking)

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Mugurell) [inline components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:831]: Desktop uses smaller URLs which are easier to read and reason about:
image
vs
image

We could decide to use a similar approach for all prompts.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt
Comment (Amejia481) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt:71]: nit: I was a bit afraid that name could be easily confused with the `ContentPersistentStorage`. I was thinking about a more descriptive name like  `CrossOriginAccessStorage` what do you think?

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [inline components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:777]: Nit: Maybe we could use a [cookie icon](https://design.firefox.com/icons/viewer/#cookie)?  :)

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:831]: Agree maybe [stripDefaultPort](https://github.com/mozilla-mobile/android-components/blob/3b2d7a30fe7a558bd6661c3152d5a611f88cac9d/components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt#L171) or [getOrigin()](https://github.com/mozilla-mobile/android-components/blob/3b2d7a30fe7a558bd6661c3152d5a611f88cac9d/components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt#L157) could help.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt
Comment (Mugurell) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt:71]: Indeed they are similar.
I went with using `..StorageAccess` as that is the name in the web API though.
Regarding renaming it, all other permissions start with `Content` so to combine this with your suggestion we'd get `ContentCrossOriginAccessStorage`. Should I change to this?

Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt:71]: Sounds perfect :)

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [inline components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:76]: nit: Maybe could allow apps to pass their own link and make this one as the default value, as apps may have their own SUMO article page.

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [inline components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:112]: nit: Maybe apps could pass a reference of the use case?

**Review by Amejia481 — COMMENTED** [review]
This looks great, before approving I'll do round of manual testing just to verify that everything is OK. Awesome work 👏🏽 .

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Mugurell) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:76]: Since the entire functionality is based on the Mozilla API and the Gecko / GVs implementation details I think it would be okay to hardcode the url. 
Also this AC feature seems like a standalone one with a general API so putting this highly specific detail in the constructor for just one permission would make it a bit more cumbersome?

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Mugurell) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:112]: Thought about this in the scenario that the the usecase is being passed in already consatructed with a specific store.
But since the feature receives the store as an argument and used it already to search for tabs or observe various changes it seems like the functionality is already tied to this specific store.
Then I thought that we could `hoist` the event of the user clicking the "Learn more" link and let the user decide what should happen - maybe use a different usecase, which would've also helped with the above scenario of letting clients control the documentation URL but this seemed like a bit too bigger change for just one permission from the many that the feature helps with.

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:76]: Good point👌🏽 . Let's leave it as it's :)

**Review by Amejia481 — APPROVED** [review]
After testing manually this works like a charm, amazing work 🥇 !
As we are close to the weekend let's wait until early next week to land.

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:112]: 👍🏽

**Review by topotropic — CHANGES_REQUESTED** [review]
Thanks for the ping!
Could you change the color of the "Learn more" link to #592ACB for light and #AB71FF for dark mode? And, where do we link to when people tap on  "Learn more"?

**Comment by Mugurell** [comment]
> Could you change the color of the "Learn more" link to #592ACB for light and #AB71FF for dark mode? 

Sure! These values will be added in Fenix - https://github.com/mozilla-mobile/fenix/pull/22855 since AC doesn't know about light/dark mode.

> And, where do we link to when people tap on "Learn more"?

To the Mozilla documentation page for this permission - https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API

**Comment by Mugurell** [comment]
Checked "offline" with Nicole on what would still be needed and looks like with the colors change done this is ready to be merged.

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (topotropic) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt]: @Mugurell what's the current copy we have on mobile for this dialog? Is it the short one "Allow .... to use cookies on ..." or is it the slightly longer on "Allow .... to use its cookies on ...? You may want to block access if it's not clear why.... "? Thanks!

Comment (Mugurell) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt]: @topotropic Tested just now. We are using the slightly longer version with [these two strings](https://github.com/mozilla-mobile/android-components/blob/8107b87e0898f99baa900217d5cd8e214cef3ec0/components/feature/sitepermissions/src/main/res/values/strings.xml#L41-L46).




&nbsp;

Seems like desktop does the same:
image

----
Will update our negative button to also say `Block` instead of `Don't allow` based on https://mozilla-hub.atlassian.net/browse/FNXV2-20626

**Pull request merged**
